src/cli/index.js

```diff
@@ -234,68 +235,129 @@ export function main({
+          // Don't forget to kill the sockets if we're being killed via signals
+          onDeath(killSockets);
+
+          // Also kill the sockets if we finish, whether it's a success or a failure
+          run().then(
+            res => {
+              killSockets();
+              resolve(res);
+            },
+            err => {
+              killSockets();
+              reject(err);
+            },
+          );
+        });
+
+        function manager(request, response) {
+          response.writeHead(200);
+          response.write(config.cwd);
+          response.end();
+        }
 
-      const onServerEnd = async () => {
-        for (const client of clients) {
+        function killSockets() {
           try {
-            client.destroy();
+            server.close();
           } catch (err) {
-            // pass
+            // best effort
+          }
+
+          for (const socket of clients) {
+            try {
+              socket.destroy();
+            } catch (err) {
+              // best effort
+            }
           }
+
+          // If the process hasn't exited in the next 5s, it has stalled and we abort
+          const timeout = setTimeout(() => {
+            console.error('Process stalled');
+            if (process._getActiveHandles) {
+              console.error('Active handles:');
+              // $FlowFixMe: getActiveHandles is undocumented, but it exists
+              for (const handle of process._getActiveHandles()) {
+                console.error(`  - ${handle.constructor.name}`);
+              }
+            }
+            // eslint-disable-next-line no-process-exit
+            process.exit(1);
+          }, 5000);
+
+          // This timeout must not prevent us from exiting
+          // $FlowFixMe: Node's setTimeout returns a Timeout, not a Number
+          timeout.unref();
         }
+      }
 
-        await server.close();
-        server.unref();
-      };
+      function reportServerName() {
+        const request = http.get(connectionOptions, response => {
+          const buffers = [];
 
-      // open the server and continue only if succeed.
-      server.listen(connectionOptions, () => {
-        // ensure the server gets closed properly on SIGNALS.
-        onDeath(onServerEnd);
+          response.on('error', () => {
+            startServer();
+          });
 
-        resolve(run().then(onServerEnd));
-      });
+          response.on('data', buffer => {
+            buffers.push(buffer);
+          });
 
-      server.on('connection', function(socket) {
-        clients.add(socket);
+          response.on('end', () => {
+            reporter.warn(reporter.lang('waitingNamedInstance', Buffer.concat(buffers).toString()));
+            waitForTheNetwork();
+          });
+        });
 
-        socket.on('close', () => clients.delete(socket));
-      });
+        request.on('error', () => {
+          startServer();
+        });
+      }
```

Comment:
Looks like you don't call `request.end()` in anywhere. According to Node docs, it should always be invoked:

https://nodejs.org/api/http.html#http_http_request_options_callback
>Note that in the example req.end() was called. With http.request() one must always call req.end() to signify the end of the request - even if there is no data being written to the request body.

Reply:
I use `http.get()`, not `http.request()`.

> The only difference between this method and `http.request()` is that it sets the method to `GET` and calls `req.end()` automatically.